pkg/tests/schema_pulumi_test.go

```diff
@@ -2592,3 +2596,360 @@ Resources:
+		{
+			name: "TypeList do not validate nested fields",
+			cloudVal: []interface{}{
+				map[string]interface{}{
+					"nested_prop":       "ABC",
+					"nested_other_prop": "ABC",
+				},
+			},
+			schema: schema.Schema{
+				Type:     schema.TypeList,
+				Required: true,
+				MaxItems: 1,
+				Elem: &schema.Resource{
+					Schema: map[string]*schema.Schema{
+						"nested_prop": {
+							Type:     schema.TypeString,
+							Optional: true,
+							ValidateDiagFunc: func(i interface{}, p cty.Path) diag.Diagnostics {
+								return diag.Errorf("Error")
+							},
+						},
+						"nested_other_prop": {
+							Type:     schema.TypeString,
+							Optional: true,
+							ValidateDiagFunc: func(i interface{}, p cty.Path) diag.Diagnostics {
+								return nil
+							},
+						},
+					},
+				},
+			},
+			expectedProps: map[string]interface{}{
+				"collectionProp": map[string]interface{}{
+					"nestedOtherProp": "ABC",
+					"nestedProp":      "ABC",
+				},
+			},
+			expectFailure: true,
+		},
+		{
+			name: "TypeSet Do not validate nested fields",
+			cloudVal: []interface{}{
+				map[string]interface{}{
+					"nested_prop": "ABC",
+				},
+			},
+			schema: schema.Schema{
+				Type:     schema.TypeSet,
+				Required: true,
+				Elem: &schema.Resource{
+					Schema: map[string]*schema.Schema{
+						"nested_prop": {
+							Type:     schema.TypeString,
+							Required: true,
+							ValidateFunc: func(i interface{}, s string) ([]string, []error) {
+								return []string{}, []error{errors.New("Error")}
+							},
+						},
+					},
+				},
+			},
+			expectedProps: map[string]interface{}{
+				"collectionProps": []interface{}{
+					map[string]interface{}{
+						"nestedProp": "ABC",
+					},
+				},
+			},
+			expectFailure: true,
+		},
+	}
+
+	for _, tc := range tests {
+		tc := tc
+		t.Run(tc.name, func(t *testing.T) {
+			resMap := map[string]*schema.Resource{
+				"prov_test": {
+					Schema: map[string]*schema.Schema{
+						"collection_prop": &tc.schema,
+						"other_prop": {
+							Type:     schema.TypeString,
+							Optional: true,
+						},
+					},
+					Importer: &schema.ResourceImporter{
+						StateContext: func(ctx context.Context, rd *schema.ResourceData, i interface{}) ([]*schema.ResourceData, error) {
+							err := rd.Set("collection_prop", tc.cloudVal)
+							assert.NoError(t, err)
+							err = rd.Set("other_prop", "test")
+							assert.NoError(t, err)
+							return []*schema.ResourceData{rd}, nil
+						},
+					},
+					ReadContext: func(ctx context.Context, rd *schema.ResourceData, i interface{}) diag.Diagnostics {
+						err := rd.Set("collection_prop", tc.cloudVal)
+						assert.NoError(t, err)
+						err = rd.Set("other_prop", "test")
+						assert.NoError(t, err)
+						return nil
+					},
+				},
+			}
+			tfp := &schema.Provider{ResourcesMap: resMap}
+			bridgedProvider := pulcheck.BridgedProvider(t, "prov", tfp, pulcheck.DisablePlanResourceChange())
```

Comment:
Brilliant, this is managing to test end to end somehow. Big step up for the validity of these tests!